Alertinator.php

```diff
@@ -68,7 +83,76 @@ public function check() {
          }
       }
    }
+   
+   /**
+    * Threshold notification: determine if an all-clear alert should be sent,
+    * and if so, send it and reset the logger.
+   */
+   private function notifyClear($check, $alertAfter, $clearAfter, $alerteeGroups) {
+      $log = $this->logger->readAlerts($check);
+      krsort($log);
+      
+      // We only get here if there was at least 1 failure, but 1 failure may not
+      // exceed the alertAfter threshold. If the check succeeds without
+      // reaching the alert threshold, reset the log silently.
+      // TODO: this will break terribly for bouncing errors, e.g. pass-> fail->
+      //       pass-> fail-> pass-> fail. Account for this.
+      if (count($log) < $alertAfter) {
+         $this->logger->resetAlerts($check);
+         return;
+      }
+      
+      if (count($log) >= $clearAfter) {
+         $clears = 0;
+         // Only notify a clear if the check passes > $clearAfter in a row.
+         foreach($log as $alert) {
+            if (!$alert['status']) {
+               break;
+            }
+            $clears++;
+         }
+         if ($clears >= $clearAfter) {
+            $last = end($log);
+            $message = "The alert '$check' was cleared at " . date(DATE_RFC2822, $last['ts']) . ".";
+            // TODO: It is impossible here to know what exception level should
+            // be sent. In absence of this information we have to just blast it
+            // to everyone. Once https://github.com/iFixit/alertinator/issues/3
+            // is implemented, exception level should be available.
+            $e = new AlertinatorCriticalException($message);
+            $this->alertGroups($e, $alerteeGroups);
+            $this->logger->resetAlerts($check);
+         }
+      }
+   }
+    
+   /**
+    * Threshold notification: determine if a failure alert should be sent, and
+    * if so, send it.
+   */  
+   private function notifyFailure($check, $alertAfter, $alerteeGroups, $e, $remindEvery) {
+      $log = $this->logger->readAlerts($check);
+      $fails = 0;
 
+      foreach ($log as $alert) {
+         if (!$alert['status']) {
+            $fails++;
+         }
+      }
+
+      $reminderTriggered = $fails > $alertAfter
+       && ($fails - $alertAfter) % $remindEvery == 0;
+
+      if ($fails && $fails >= $alertAfter && ($fails === $alertAfter || $reminderTriggered)) {
```

Comment:
This could be $fails && ($fails === $alertAfter || $reminderTriggered) I think, since $reminderTriggered alaready checks $fails > $alertAfter